x/expiration/types/codec.go

```diff
@@ -0,0 +1,43 @@
+package types
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/msgservice"
+
+	metadatatypes "github.com/provenance-io/provenance/x/metadata/types"
+)
+
+// RegisterInterfaces registers the module 'Msg' types with the registry
+func RegisterInterfaces(registry types.InterfaceRegistry) {
+	registry.RegisterImplementations((*sdk.Msg)(nil),
+		&MsgAddExpirationRequest{},
+		&MsgExtendExpirationRequest{},
+		&MsgInvokeExpirationRequest{},
+	)
+
+	msgservice.RegisterMsgServiceDesc(registry, &_Msg_serviceDesc)
+}
+
+// whiteListedInterfaceRegistry whitelist of sdk.Msg types
+// invoked when a module asset expiration has expired
+func whiteListedInterfaceRegistry() types.InterfaceRegistry {
+	ir := types.NewInterfaceRegistry()
+	ir.RegisterImplementations((*sdk.Msg)(nil),
+		&metadatatypes.MsgDeleteScopeRequest{},
+		&metadatatypes.MsgDeleteRecordRequest{},
```

Comment:
I'd like the expiration module to not have anything referencing modules that will be using it. It'd be pretty easy to end up with circular references. Maybe that's something to address later though.

Reply:
Good catch! I had this in place when we were supporting `add` expiration on the cli to be able to run `ValidateBasic()` on the message that would be invoked during a tx broadcast (it would result in an error with ModuleCdc not having those registered). We have since removed the `add` functionality from the cli and these can be removed. However, we should restrict storing only `delete` message types in an expiration record. Do you have a recommendation on a way to do that without having to reference other modules?

Reply:
I think the way to do this is to have a field and registration function to the expiration keeper that stores and adds-to such a whitelist. Then, in a module that is going to use the expiration module (e.g. `x/metadata/module.go`), create a function that takes in an expiration keeper and registers the allowed messages. It wouldn't allow for use during basic validation, but could still be enforced in the other validation.

Then, in `app/app.go`, we'd add a function `initExpirationKeeper` (and call it in `New`) that calls all the module functions that register the whitelisted messages. The easy way to do it is to just hard-code all needed calls (similar to `initParamsKeeper`). The fun way would be to define an interface that a module would satisfy if they want to register whitelisted messages, then go through each module in the module manager and (if it implements that interface), call the method.

Reply:
👍 I'm going to attack this in a new ticket -> #1132.